File: backend/schemas/markets.py
from typing import Optional

from pydantic import BaseModel


class SymbolInfo(BaseModel):
    symbol: str
    baseAsset: str
    quoteAsset: str
    stepSize: Optional[str] = None      # LOT_SIZE filter
    tickSize: Optional[str] = None      # PRICE_FILTER filter
    minNotional: Optional[str] = None   # MIN_NOTIONAL filter


class MarketStats(BaseModel):
    symbol: str
    lastPrice: str
    priceChange: str
    priceChangePercent: str
    highPrice: str
    lowPrice: str
    openPrice: str
    prevClosePrice: str
    bidPrice: str
    askPrice: str
    volume: str
    quoteVolume: str
    openTime: int
    closeTime: int


class Candle(BaseModel):
    time: int    # kline open time in ms
    open: str
    high: str
    low: str
    close: str
    volume: str


class PriceHistory(BaseModel):
    symbol: str
    period: str
    candles: list[Candle]

File: frontend/src/api/markets.ts
import apiClient from "./client";

export interface TradingSymbol {
  symbol: string;
  baseAsset: string;
  quoteAsset: string;
  stepSize: string;
  tickSize: string;
  minNotional: string;
}

export interface MarketData {
  symbol: string;
  lastPrice: string;
  priceChange: string;
  priceChangePercent: string;
  highPrice: string;
  lowPrice: string;
  volume: string;
  quoteVolume: string;
  openPrice: string;
  prevClosePrice: string;
  bidPrice: string;
  askPrice: string;
  openTime: number;
  closeTime: number;
}

export interface Candle {
  time: number;
  open: string;
  high: string;
  low: string;
  close: string;
  volume: string;
}

export interface PriceHistory {
  symbol: string;
  period: string;
  candles: Candle[];
}

export async function getMarketData(symbol: string): Promise<MarketData> {
  const { data } = await apiClient.get<MarketData>(`/markets/${symbol}`);
  return data;
}

export async function getPriceHistory(symbol: string, period: string): Promise<PriceHistory> {
  const { data } = await apiClient.get<PriceHistory>(`/markets/${symbol}/history`, {
    params: { period },
  });
  return data;
}

export async function getSymbols(): Promise<TradingSymbol[]> {
  const { data } = await apiClient.get<TradingSymbol[]>("/symbols");
  return data;
}
